doc/proposals/webhook-proposal.md

```diff
@@ -0,0 +1,137 @@
+```go
+type webhookServer struct {
+    Port int
+    CertDir string
+} 
+```
+
+The initialization function will be called in the main.go file, regardless of whether or not a user opts in, however no actions are taken unless the user chooses to generate webhooks.
+
+```go
+func WebhookInit() error {
+    log.Info("Starting webhook server")
+    hookServer := &webhook.Server{
+        Port: hookServerCfg.Port,
+        CertDir: hookServerCfg.CertDir,
+    }
+    
+    if err := mgr.Add(hookServer); err != nil {
+        log.Error(err, "Unable to register webhook server")
+        return err
+    }
+
+    log.Info("Registering webhooks to the webhook server")
+    
+    hookServer.Register("/mutate-pods", &webhook.Admission{Handler: &podAnnotator{}})
+    hookServer.Register("/validate-pods", &webhook.Admission{Handler: &podValidator{}})
+    
+    return nil
+}
+```
+
+
+In addition to the above file being used to start up the server, an additional  files (`mutationswebhook.go` and `validationwebhooks.go`) will generated when the user chooses to include webhooks. These are the files which have the necesary boilerplate for a user to implement the `Handle` function in order to run custom logic on validation and admission of pods.
+
+```go
+type podInteraction struct {
+	client  client.Client
+	decoder *admission.Decoder
+}
+
+func (i *podInteraction) Handle(ctx context.Context, req admission.Request) admission.Response {
+	pod := &corev1.Pod{}
+
+	err := i.decoder.Decode(req, pod)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+    //Insert mutation/validation logic here
+
+    //BRANCH
+    //Block A contains the boiler plate logic for patching a mutation
+    //Block B contains the boilder plate logic for 
+
+    //BLOCK A
+	marshaledPod, err := json.Marshal(pod)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+    return admission.PatchResponseFromRaw(req.Object.Raw, marshaledPod)
+    
+    //BLOCK B
+    if FAILURE_CONDITION {
+		return admission.Denied(fmt.Sprintf("reason for failure"))
+	}
+
+    return admission.Allowed("")
+    
+    //END BRANCH
+}
+
+// podInteraction implements inject.Client.
+// A client will be automatically injected.
+
+// InjectClient injects the client.
+func (i *podInteraction) InjectClient(c client.Client) error {
+	i.client = c
+	return nil
+}
+
+// podInteraction implements admission.DecoderInjector.
+// A decoder will be automatically injected.
+
+// InjectDecoder injects the decoder.
+func (i *podInteraction) InjectDecoder(d *admission.Decoder) error {
+	i.decoder = d
+	return nil
+}
+
+```
+
+
+### User facing usage (if needed)
+
+My suggested method for interacting with this feature is to have a command in the osdk which can be run after generating the base operator. The new command `$ operator-sdk generate webhook` will write the files cmd/manager/osdk-webhook.go, cmd/manager/mutationwebhook.go, and cmd/manager/validationwebhook.go
```

Comment:
```suggestion
My suggested method for interacting with this feature is to have a command in the osdk which can be run after generating the base operator. The new command `$ operator-sdk generate webhook` will write the files `cmd/manager/osdk-webhook.go`, `cmd/manager/mutationwebhook.go`, and `cmd/manager/validationwebhook.go`.
```

Reply:
@theishshah For operators with multiple APIs, we need to be able to support multiple webhooks. Consider the following scenario:

1. User runs `operator-sdk new my-operator`

    Right now, this scaffolds out a `main.go` and some boilerplate in `pkg/apis` and `pkg/controllers` for future calls to `operator-sdk add api` and `operator-sdk add controller`, respectively.

2. User runs `operator-sdk up local`

    This will compile and run successfully but nothing happens since we have no controllers or APIs

3. User runs:
    -  `operator-sdk add api --api-verison=example.com/v1alpha1 --kind MyDatabase`
    -  `operator-sdk add api --api-verison=example.com/v1alpha1 --kind MyWebApp`
    -  `operator-sdk add controller --api-verison=example.com/v1alpha1 --kind MyDatabase`
    -  `operator-sdk add controller --api-verison=example.com/v1alpha1 --kind MyWebApp`

    None of these commands modify anything in `cmd/manager`. Instead the scaffolded files register themselves via the functions created previously by `operator-sdk new` in `pkg/apis` and `pkg/controller`. Now, when running `operator-sdk up local`, these APIs and controllers are picked up and CRs are reconciled.

4. Now the user wants to add some webhooks. Going along with this pattern, that probably means something like a `operator-sdk add webhook` command that takes:

    - `--api-version` and `--kind` so that we know where to create the file(s) that define the webhooks for a specific type.
    - `--webhook-type <validating|mutating|conversion>` so that we know which types of webhooks to create (maybe we have one file per type?)


5. The user runs the following to create two webhooks:
    - `operator-sdk add webhook --api-verison=example.com/v1alpha1 --kind MyDatabase --webhook-type=mutating`
    - `operator-sdk add webhook --api-verison=example.com/v1alpha1 --kind MyWebApp --webhook-type=validating`

    These, perhaps, would end up creating:
    - `pkg/apis/example/v1alpha1/mydatabase_mutating_webhook.go`
    - `pkg/apis/example/v1alpha1/mywebapp_validating_webhook.go`
 

As discussed offline, it would be really helpful to include links in the proposal that contain examples, types, functions, etc. provided by controller-runtime that we intend to use. That would give reviewers more context about the proposal.